File: Exercism Codes/JavaScript/matching-brackets.js
export const isPaired = (text) => {
  const stack = [];

  for (const x of text) {
    if ("([{".includes(x)) {
      stack.push(x);
    }
    if (x === ")") {
      if (stack.length > 0 && stack[stack.length - 1] === "(") {
        stack.pop();
      }
      else return false;
    }
    if (x === "]") {
      if (stack.length > 0 && stack[stack.length - 1] === "[") {
        stack.pop();
      }
      else return false;
    }
    if (x === "}") {
      if (stack.length > 0 && stack[stack.length - 1] === "{") {
        stack.pop();
      }
      else return false;
    }
  }
  
  return stack.length === 0;
};

File: Exercism Codes/Python/matching_brackets.py
def is_paired(input_string):
    balance_stack = []

    for c in input_string:
        if c in "([{":
            balance_stack.append(c)
        if c == ')':
            if len(balance_stack) > 0 and balance_stack[-1] == '(':
                balance_stack.pop()
            else:
                return False
        if c == ']':
            if len(balance_stack) > 0 and balance_stack[-1] == '[':
                balance_stack.pop()
            else:
                return False
        if c == '}':
            if len(balance_stack) > 0 and balance_stack[-1] == '{':
                balance_stack.pop()
            else:
                return False

    return len(balance_stack) == 0
